### PR TITLE
Differentiate is_int primitive between generic and variant-only versions

### PR DESCRIPTION
**Pull request by lthls**
This is a proper version of #747.
The `Pisint` lambda primitive and `Is_int` flambda2 primitive are augmented with a boolean distinguishing the generic version, exposed as `Obj.is_int`, and the version used for compiling pattern-matching on variants.
This allows Flambda 2 to still precisely handle variants without being incorrect on the `Obj.is_int` version.

### CODE REVIEW

**Comment by xclerc** [comment]
I was not able to get to the bottom of the `fmt` issue but
I suspect the semantics of `**` is at play. Could you try to
manually expand the `**` in the various ocamlformat files
and see whether that makes a difference? (it has, for me,
on another PR)

**Comment by lthls** [comment]
> Could you try to
> manually expand the `**` in the various ocamlformat files
> and see whether that makes a difference? (it has, for me,
> on another PR)

It seems to work: after expanding the `**` in `middle_end/.ocamlformat-enable`, my local `make fmt` produces the same result as the CI.
The manual expansion doesn't look great though, not I'm not sure I should commit it.

**Comment by xclerc** [comment]
I am not necessarily saying the expansion should be
committed, but it looks like either a bug in ocamlformat
or a misinterpretation of the intended semantics on our
part. In both cases, something we would like to discuss
with the ocamlformat developers; thanks for checking.

**Comment by xclerc** [comment]
@gpetiot has figured the cause: `ocaml-re` 1.9.0 does
not work as expected, while 1.10.x does.

**Comment by lthls** [comment]
Thanks, I had `re.1.9.0` installed indeed. I've upgraded now, so hopefully I shouldn't see these errors anymore.

File: middle_end/flambda2/simplify/simplify_unary_primitive.ml [hunk 1]
```diff
@@ -180,10 +180,23 @@ let simplify_is_int_or_get_tag dacc ~original_term ~scrutinee ~scrutinee_ty:_
...
+      ~result_var ~make_shape:(fun scrutinee ->
+        T.is_int_for_scrutinee ~scrutinee)
+  else
+    match T.prove_is_int (DA.typing_env dacc) scrutinee_ty with
+    | Proved b ->
+      let ty = T.this_naked_immediate (Targetint_31_63.bool b) in
+      let dacc = DA.add_variable dacc result_var ty in
+      SPR.create original_term ~try_reify:false dacc
+    | Unknown ->
+      let ty = T.unknown K.naked_immediate in
```
Comment (mshinwell) [inline middle_end/flambda2/simplify/simplify_unary_primitive.ml:197]: This should probably use `SPR.create_unknown`.

**Review by mshinwell — CHANGES_REQUESTED** [review]

File: middle_end/flambda2/simplify/simplify_unary_primitive.ml (hunk 1 above)
Comment (lthls) [reply middle_end/flambda2/simplify/simplify_unary_primitive.ml]: I've fixed it, along with the remaining similar cases in the same file.

**Review by mshinwell — APPROVED** [review]

**Pull request merged**